### ISSUE TITLE
Generate OCaml modules containing ood data

### ISSUE THREAD

**Issue by tmattio** [ocaml/ood#35]
Currently, `ood` only provides types for the different data stored in the repository. This has a few consequences:

- The data has to be read and processed by the user. This is particularly problematic as we are using two different processors: `omd` in `ood-preview` to test that everything is ok, and another JS library (?) in `v3.ocaml.org`. This deprecancy opens the door for inconsistencies between the data that has been QAed during the import, and the one available in `v3`.
- Errors in the data are only caught at runtime.
- The added complexity in user-site to work with the different processors and filesystem is huge

I propose that we write a preprocessor that will output OCaml modules containing the data available in `data/`.
The output module will not have any dependency, so it will be usable, regardless of the setup, and the data will be consistent with all of the projects that use `ood`.

**Comment by patricoferris** [comment]
Interesting idea, to check that I fully understand it, would this be like doing an `ocaml-crunch` of files except down to the individual meta-data level storing each file as an OCaml value? So for example, a basic tutorial structure might be: 

```ocaml
type metadata = {
  title : string;
  description : string;
  authors : string list; 
}
```

And suppose we have tutorials `basic.md` and `advanced.md`, is the proposal to automatically generate and promote a file called `tutorials.ml` with (vaguely) the following: 

```ocaml
type metadata = {
  title : string;
  description : string;
  authors : string list; 
}

type t = {
  meta : metadata;
  body : string;
}

let basic = { 
  meta : {
    title = "basic";
    description = "a basic tutorial";
    authors = [ "Alice"; "Bob" ]; 
  };
  body = "## My tutorial\nblah blah blah"
}

(* similarly for advanced *)

let tutorials = [ basic; advanced ]
```

This can then be consumed by others?

**Comment by tmattio** [comment]
Yes, exactly! With the addition that the markdown preprocessing would be done at generation time as well, so we could provide a field `body_md` and `body_html`.

**Comment by tmattio** [comment]
This would also be very useful to encode some assumptions we take on the structure of the makdown. For instance, in #32, we assume there is only one `Statement` and `Solution` heading. This assumption is hard to share and keep consistent with user code, but if we generate the data ourselves, we can simply output the two fields `statement_md` and `solution_md`.

**Comment by patricoferris** [comment]
Hmm this could work. From my limited understanding I feel like this is kind of the equivalent of packaging up data in an npm package except this would be typed. This might also greatly simplify (as you pointed out) consuming the data because all of the parsing, linting and data-checking is centralised to ood and if a type changes you won't get a runtime parsing error elsewhere, but instead a compiler type error if I'm not mistaken. @avsm, @agarwal, @kanishka-work and @rdavison any thoughts from your end on this?

**Comment by agarwal** [comment]
In https://github.com/ocaml/ood/pull/36#issuecomment-858070924, I said this could be "useful". However, I realize we discussed only that it could be "feasible". I'm not sure yet of the benefit. This adds a layer of engineering and indirection. We should probably articulate the value before committing to this.

**Comment by tmattio** [comment]
Thanks @agarwal!

I tried to articulate the problems of the current approach and the benefits of this new approach in this issue and the PR.

The immediate benefits I can see:

- Remove complexity from the frontend by dealing with parsing and processing in ood, so it can stay focused on frontend-related work.
- Ensure the data is consistent with `ood-preview`, which @patricoferris and I use to make sure our data is ready to be used in `v3`.
- Catch data errors at compile time.

There's also the fact, as we discussed, that processing the data should not be a concern of the frontend, and unless we put some more business logic into ood, we'll have to tackle typical backend work on the ReScript side, which is not adapted for this and will only slow us down during the implementation of the design.

I can't see any drawback of this approach, so given the above, it seemed natural to me to go with it. I thought we reached a consensus on this, but apparently not, so I apologize for moving forward too quickly on this. I'm very open to any insights/concerns you think we should discuss first 🙂

**Comment by tmattio** [comment]
> This adds a layer of engineering and indirection

To answer this specifically, I don't think this is true. The layer added to ood aims at replacing the one in ReScript, so this is shifting, rather than adding, a layer of engineering.

And after working on removing the markdown and Yaml processing from the frontend to use the static modules, I can also tell that the complexity of the processing is far greater in the frontend at the moment when compared to processing done w/ Omd in #36, so I think we're actually reducing the overall complexity with this new approach (koodos to Omd and `ppx_deriving_yaml` for this, they are great libraries).

This might sound a bit abstract for now, as I didn't land a PR yet, but I hope the gains on the overall complexity will become clearer once I'm able to shift my focus off from ood to v3 and propose a PR to migrate the codebase!

**Comment by avsm** [comment]
In #36, I noted this which is quite important:
> The advantage of markdown processing in omd is that we use a single implementation instead of two. There are lots of subtle variations and extensions in implementations, but at least with omd we can extend it to our needs. We can certainly modify omd2 fairly easily to account for the specific html needed by v3.ocaml.org. (and in general, ensure omd is compatible with tailwind 2)

It's really important we don't have multiple ways of interpreting our site-data-at-rest (which is all in ood now). I can see all sorts of subtle rendering errors coming to the fore due to different implementations of Markdown handling things differently.  With a clean back/front split that we have with ood/v3, the interpretation of the authoring format (markdown especially) seems better to live in the backend, and the rendering logic in the frontend.  I agree with @tmattio that this is removing layers, since there are strictly fewer pieces of code interpreting a bit of data. From a maintenance perspective, we can evolve `omd` to satisfy all of our Markdown authoring needs into the long term.

More pragmatically, I think we need to take a decision on this early and not let the PR changes pile up in `ood/next`, as it's a big set of changes that'll start to conflict.

**Comment by agarwal** [comment]
All sounds good! Thanks for the explanations.

**Comment by tmattio** [comment]
This has been fixed in #36
